FV3/gfsphysics/GFS_layer/GFS_physics_driver.F90

```diff
@@ -5526,7 +5561,196 @@ subroutine moist_bud2(im,ix,ix2,levs,me,kdt,grav,dtp,delp,rain, &
+    ! air, and emit a warning.
+#ifdef MULTI_GASES
+    call mpp_error (NOTE, 'GFS_physics_driver::moist_cp - moist_cp for tracer configuration not implemented; using default cp_air for t_dt diagnostics')
+    cpm = cp_air
+#else
+    if (nwat /= 6) then
+      call mpp_error (NOTE, 'GFS_physics_driver::moist_cp - moist_cp for tracer configuration not implemented; using default cp_air for t_dt diagnostics')
+      cpm = cp_air
+    else
+      call physics_to_dycore_mass_fraction(initial_dynamics_q, physics_q, &
+          pressure_on_interfaces, im, levs, nwat, new_dynamics_q)
+
+      q_vapor = new_dynamics_q(:,:,ntqv)
+      q_liquid = new_dynamics_q(:,:,ntcw) + new_dynamics_q(:,:,ntrw)
+      q_ice = new_dynamics_q(:,:,ntiw) + new_dynamics_q(:,:,ntsw) + new_dynamics_q(:,:,ntgl)
+      q_dry_air = 1.0 - q_vapor - q_liquid - q_ice
+
+      ! By definition now, the weights sum to 1.0.
+      cpm = cp_air * q_dry_air + cp_vap * q_vapor + c_liq * q_liquid + c_ice * q_ice
+    endif
+#endif
+  end subroutine moist_cp
+
+      subroutine physics_to_dycore_mass_fraction(initial_dynamics_q, physics_q, &
+            pressure_on_interfaces, im, levs, nwat, new_dynamics_q)
+        integer, intent(in) :: im, levs, nwat
+        real(kind=kind_phys), intent(in) :: initial_dynamics_q(1:im,1:levs,1:nwat)
+        real(kind=kind_phys), intent(in) :: physics_q(1:im,1:levs,1:nwat)
+        real(kind=kind_phys), intent(in) :: pressure_on_interfaces(1:im,1:levs+1)
+        real(kind=kind_phys), intent(out) :: new_dynamics_q(1:im,1:levs,1:nwat)
+
+        real(kind=kind_phys) :: delp(1:im,1:levs)
+        real(kind=kind_phys) :: qwat(1:im,1:levs,1:nwat)
+        real(kind=kind_phys) :: qt(1:im,1:levs)
+        integer :: k, t
+
+        ! Follow the procedure used in atmosphere.atmosphere_state_update to convert
+        ! mass fractions in the physics driver to mass fractions in the dynamical core.
+        do k = 1, levs
+          delp(:,k) = pressure_on_interfaces(:,k) - pressure_on_interfaces(:,k+1)
+        enddo
+        do t = 1, nwat
+          qwat(:,:,t) = delp * physics_q(:,:,t)
+        enddo
+        qt = sum(qwat, 3)
+        delp = delp * (1.0 - sum(initial_dynamics_q(:,:,1:nwat), 3)) + qt
+        do t = 1, nwat
+          new_dynamics_q(:,:,t) = qwat(:,:,t) / delp
+        enddo
+      end subroutine physics_to_dycore_mass_fraction
+
+      subroutine infer_nwat(ntqv, ntcw, ntiw, ntrw, ntsw, ntgl, nwat)
+        integer, intent(in) :: ntqv, ntcw, ntiw, ntrw, ntsw, ntgl
+        integer, intent(out) :: nwat
+
+        integer, dimension(6) :: tracer_indexes
+        integer, dimension(6) :: tracers_active
+
+        tracer_indexes = (/ ntqv, ntcw, ntiw, ntrw, ntsw, ntgl /)
+
+        where (tracer_indexes > 0)
+          tracers_active = 1
+        elsewhere
+          tracers_active = 0
+        endwhere
+        nwat = sum(tracers_active)
+      end subroutine infer_nwat
```

Comment:
I thought nwat was a namelist variable, why does it need to be inferred?

Reply:
Indeed, it is in the dynamical core, but not in the physics.  I *could* pass it through the `Statein`, similar to the `hydrostatic` flag.  Maybe I should just do that...